File: java/src/test/java/com/benchmark/PhysicsTest.java
package com.benchmark;

import com.benchmark.models.Config;
import com.benchmark.models.Edge;
import com.benchmark.models.Vehicle;
import com.benchmark.physics.Physics;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit verification suite for vehicle physics and aerodynamic calculations.
 *
 * <p>Validates each physical sub-formula independently against analytically expected
 * values derived from first-principles mechanics. Uses a consistent 1,400 kg
 * medium passenger vehicle specification matching the canonical benchmark dataset.
 */
public class PhysicsTest {

    private Vehicle vehicle;
    private Config config;

    /**
     * Initialises shared vehicle and configuration fixtures before each test method.
     */
    @BeforeEach
    public void setUp() {
        vehicle = new Vehicle(1400.0, 0.30, 2.2, 0.012, 0.30, 34.2e6, 40.0);
        Map<String, Double> speedFactors = new HashMap<>();
        speedFactors.put("HIGHWAY",  1.05);
        speedFactors.put("MAIN_ROAD", 0.95);
        config = new Config(1.225, 9.81, 0.40, 0.40, 0.20, speedFactors);
    }

    /**
     * Drag force must equal 0.5 * rho * Cd * A * v^2 in zero-wind conditions.
     */
    @Test
    @DisplayName("Aerodynamic drag matches F = 0.5 * rho * Cd * A * v^2 at zero wind")
    public void testAerodynamicDrag() {
        Edge edge = new Edge();
        edge.distance      = 10.0;
        edge.speedLimit    = 100.0;
        edge.averageSpeed  = 100.0;
        edge.roadType      = "HIGHWAY";
        edge.trafficFactor = 1.0;

        double vMs     = 100.0 / 3.6;
        double drag    = Physics.calculateAerodynamicDrag(vMs, edge, vehicle, config);
        double expected = 0.5 * 1.225 * 0.30 * 2.2 * (vMs * vMs);
        assertEquals(expected, drag, 1e-4,
            "Aerodynamic drag deviates from analytical formula F = 0.5*rho*Cd*A*v^2.");
    }

    /**
     * Slope force at 5° incline must match m * g * sin(5°).
     */
    @Test
    @DisplayName("Slope force matches F = m * g * sin(theta) at 5-degree incline")
    public void testSlopeForce() {
        double slopeAngle = 5.0 * (Math.PI / 180.0);
        Edge edge = new Edge();
        edge.distance     = 1.0;
        edge.speedLimit   = 60.0;
        edge.averageSpeed = 60.0;
        edge.roadType     = "MAIN_ROAD";
        edge.slope        = slopeAngle;

        double fSlope   = Physics.calculateSlopeForce(edge, vehicle, config);
        double expected = 1400.0 * 9.81 * Math.sin(slopeAngle);
        assertEquals(expected, fSlope, 1e-3,
            "Slope force deviates from m * g * sin(theta).");
    }

    /**
     * Rolling resistance on flat ground must equal Crr * m * g.
     */
    @Test
    @DisplayName("Rolling resistance matches F = Crr * m * g on flat road")
    public void testRollingResistance() {
        Edge edge = new Edge();
        edge.distance     = 1.0;
        edge.speedLimit   = 60.0;
        edge.averageSpeed = 60.0;
        edge.roadType     = "MAIN_ROAD";
        edge.slope        = 0.0;

        double fRoll    = Physics.calculateRollingResistance(edge, vehicle, config);
        double expected = 0.012 * 1400.0 * 9.81 * 1.0;
        assertEquals(expected, fRoll, 1e-3,
            "Rolling resistance deviates from Crr * m * g * cos(theta).");
    }

    /**
     * Verifies that costFastest equals travelTimeMin and costCheapest equals fuel + toll.
     */
    @Test
    @DisplayName("Edge cost fields are internally consistent after computeEdgeMetrics")
    public void testEdgeMetricsConsistency() {
        Edge edge = new Edge();
        edge.distance      = 5.0;
        edge.speedLimit    = 120.0;
        edge.averageSpeed  = 110.0;
        edge.roadType      = "HIGHWAY";
        edge.slope         = 0.01;
        edge.trafficFactor = 0.9;
        edge.windSpeed     = 5.0;
        edge.tollCost      = 15.0;

        Physics.computeEdgeMetrics(edge, vehicle, config);
        assertTrue(edge.travelTimeMin > 0,
            "Travel time must be positive for a non-zero distance edge.");
        assertTrue(edge.fuelLiters > 0,
            "Fuel consumption must be positive for any traversed edge.");
        assertEquals(edge.travelTimeMin, edge.costFastest, 1e-9,
            "FASTEST cost must equal travel time in minutes.");
        assertEquals(edge.fuelCost + 15.0, edge.costCheapest, 1e-5,
            "CHEAPEST cost must equal fuel_cost + toll_cost.");
    }
}

File: go/pkg/physics/physics_test.go
package physics

import (
	"math"
	"route_optimization_go/pkg/models"
	"testing"
)

func getTestVehicle() models.Vehicle {
	return models.Vehicle{
		VehicleMass:                  1400.0,
		DragCoefficient:              0.30,
		FrontalArea:                  2.2,
		RollingResistanceCoefficient: 0.012,
		EngineEfficiency:             0.30,
		FuelEnergyDensity:            34.2e6,
		FuelPrice:                    40.0,
	}
}

func getTestConfig() models.Config {
	return models.Config{
		AirDensity:     1.225,
		Gravity:        9.81,
		FuelCostWeight: 0.40,
		TimeWeight:     0.40,
		TollWeight:     0.20,
		RoadSpeedFactors: map[string]float64{
			"HIGHWAY":   1.05,
			"MAIN_ROAD": 0.95,
		},
	}
}

func TestAerodynamicDrag(t *testing.T) {
	v := getTestVehicle()
	c := getTestConfig()
	edge := models.Edge{
		Distance:     10.0,
		SpeedLimit:   100.0,
		AverageSpeed: 100.0,
		RoadType:     "HIGHWAY",
		TrafficFactor: 1.0,
	}
	vMs := 100.0 / 3.6
	drag := CalculateAerodynamicDrag(vMs, &edge, &v, &c)
	expected := 0.5 * 1.225 * 0.30 * 2.2 * (vMs * vMs)
	if math.Abs(drag-expected) > 1e-4 {
		t.Fatalf("Drag mismatch: got %f, expected %f", drag, expected)
	}
}

func TestSlopeForce(t *testing.T) {
	v := getTestVehicle()
	c := getTestConfig()
	slopeAngle := 5.0 * (math.Pi / 180.0)
	edge := models.Edge{
		Distance:     1.0,
		SpeedLimit:   60.0,
		AverageSpeed: 60.0,
		RoadType:     "MAIN_ROAD",
		Slope:        slopeAngle,
	}
	fSlope := CalculateSlopeForce(&edge, &v, &c)
	expected := 1400.0 * 9.81 * math.Sin(slopeAngle)
	if math.Abs(fSlope-expected) > 1e-3 {
		t.Fatalf("Slope force mismatch: got %f, expected %f", fSlope, expected)
	}
}

func TestRollingResistance(t *testing.T) {
	v := getTestVehicle()
	c := getTestConfig()
	edge := models.Edge{
		Distance:     1.0,
		SpeedLimit:   60.0,
		AverageSpeed: 60.0,
		RoadType:     "MAIN_ROAD",
		Slope:        0.0,
	}
	fRoll := CalculateRollingResistance(&edge, &v, &c)
	expected := 0.012 * 1400.0 * 9.81 * 1.0
	if math.Abs(fRoll-expected) > 1e-3 {
		t.Fatalf("Rolling resistance mismatch: got %f, expected %f", fRoll, expected)
	}
}

func TestEdgeMetricsConsistency(t *testing.T) {
	v := getTestVehicle()
	c := getTestConfig()
	edge := models.Edge{
		Distance:      5.0,
		SpeedLimit:    120.0,
		AverageSpeed:  110.0,
		RoadType:      "HIGHWAY",
		Slope:         0.01,
		TrafficFactor: 0.9,
		WindSpeed:     5.0,
		TollCost:      15.0,
	}
	ComputeEdgeMetrics(&edge, &v, &c)
	if edge.TravelTimeMin <= 0 || edge.FuelLiters <= 0 {
		t.Fatalf("Invalid travel time or fuel liters: %v", edge)
	}
	if edge.CostFastest != edge.TravelTimeMin {
		t.Fatalf("Cost fastest mismatch")
	}
	if math.Abs(edge.CostCheapest-(edge.FuelCost+15.0)) > 1e-5 {
		t.Fatalf("Cost cheapest mismatch")
	}
}
